src/node_api.cc

```diff
@@ -3570,3 +3574,435 @@ napi_status napi_run_script(napi_env env,
+
+  // These are variables set once, upon creation, and then never again, which
+  // means we don't need the mutex to read them.
+  void* context;
+  size_t max_queue_size;
+
+  // These are variables accessed only from the loop thread.
+  v8impl::Reference* ref;
+  napi_env env;
+  void* thread_finalize_data;
+  napi_finalize thread_finalize_callback;
+  bool idle_running;
+  napi_async_context async_context;
+  napi_threadsafe_function_call_js call_js_cb;
+
+  // These are helpers.
+  inline node::Environment* NodeEnv() {
+    // Need a handle scope here, because this version of
+    // node::Environment::GetCurrent() establishes the current context, which
+    // needs a scope in which to live.
+    v8::HandleScope scope(env->isolate);
+    return node::Environment::GetCurrent(env->isolate);
+  }
+
+  inline void CloseHandlesAndMaybeDelete() {
+    // It is important that we delete the async first, because the async
+    // callback is the only place from which we awaken the idler. We must avoid
+    // waking the idler because we will have partially destroyed the thread-safe
+    // function by the time the idle handler runs. After all, we have already
+    // destroyed the mutex, the reference, and potentially the condition in the
+    // code above.
+    NodeEnv()->CloseHandle(
+        reinterpret_cast<uv_handle_t*>(&async),
+        [] (uv_handle_t* handle) -> void {
+          napi_threadsafe_function ts_fn =
+              node::ContainerOf(&napi_threadsafe_function__::async,
+                reinterpret_cast<uv_async_t*>(handle));
+          ts_fn->NodeEnv()->CloseHandle(
+              reinterpret_cast<uv_handle_t*>(&ts_fn->idle),
+              [] (uv_handle_t* handle) -> void {
+                bool delete_function = false;
+                napi_threadsafe_function ts_fn =
+                    node::ContainerOf(&napi_threadsafe_function__::idle,
+                      reinterpret_cast<uv_idle_t*>(handle));
+                if (ts_fn->thread_finalize_callback) {
+                  v8::HandleScope handle_scope(ts_fn->env->isolate);
+                  node::CallbackScope(ts_fn->env->isolate,
+                      v8::Object::New(ts_fn->env->isolate),
+                      *reinterpret_cast<node::async_context*>(
+                          ts_fn->async_context));
+                  ts_fn->thread_finalize_callback(ts_fn->env,
+                      ts_fn->thread_finalize_data, ts_fn->context);
+                }
+                // Destroying an async context requires a handle scope.
+                v8::HandleScope scope(ts_fn->env->isolate);
+                CHECK(napi_async_destroy(ts_fn->env, ts_fn->async_context) ==
+                    napi_ok);
+                do {
+                  node::Mutex::ScopedLock lock(ts_fn->mutex);
+                  ts_fn->handles_closed = true;
+                  if (ts_fn->thread_count == 0) {
+                    delete_function = true;
+                  }
+                } while (0);
```

Comment:
No need to `do { … } while(0)`, a block scope suffices (and doesn’t look like a loop)